src/network/networkd-dhcp4.c

```diff
@@ -216,13 +249,34 @@ static int dhcp_remove_routes(Link *link, struct in_addr *address) {
                 route->table = table;
                 route->scope = route_scope_from_address(route, address);
 
+                if (new_lease) {
+                        /* When the route is also required by new lease, then keep it.*/
+                        for (j = 0; j < m; j++) {
+                                union in_addr_union a;
+                                unsigned char l;
+
+                                assert_se(sd_dhcp_route_get_gateway(new_routes[j], &a.in) >= 0);
+                                if (!in_addr_equal(AF_INET, &a, &route->gw))
+                                        continue;
+                                assert_se(sd_dhcp_route_get_destination(new_routes[j], &a.in) >= 0);
+                                if (!in_addr_equal(AF_INET, &a, &route->dst))
+                                        continue;
+                                assert_se(sd_dhcp_route_get_destination_prefix_length(new_routes[i], &l) >= 0);
+                                if (l != route->dst_prefixlen)
+                                        continue;
+                                break;
+                        }
+                        if (j != m)
+                                continue;
+                }
+
```

Comment:
It'd be much easier to read when split out to a separate function
(also fixes j↔i typo):
```C
static bool route_present_in_routes(const Route *route, const sd_dhcp_route **routes, unsigned n_routes) {
    assert(n == 0 || routes);
    for (unsigned j = 0; j < n_routes; j++) {
                                union in_addr_union a;
                                unsigned char l;

                                assert_se(sd_dhcp_route_get_gateway(routes[j], &a.in) >= 0);
                                if (!in_addr_equal(AF_INET, &a, &route->gw))
                                        continue;
                                assert_se(sd_dhcp_route_get_destination(new_routes[j], &a.in) >= 0);
                                if (!in_addr_equal(AF_INET, &a, &route->dst))
                                        continue;
                                assert_se(sd_dhcp_route_get_destination_prefix_length(new_routes[j], &l) >= 0);
                                if (l != route->dst_prefixlen)
                                        continue;
                                return true;
                        }
         return false;
}

...

 if (route_present_in_routes(route, new_routes, m))
          continue;
```

Reply:
Thanks.

> (also fixes j↔i typo):

ouch!